### PR TITLE
[cmake] deprecate caffe2_* specific cuda function in cmake.

### PR DESCRIPTION
**Pull request by Yangqing**
Was chatting with @pietern about the cuda situation and custom paths, and this is a quick experiment to see how we can consolidate things. In the past, we have CUDA_ARCH_NAME and TORCH_CUDA_ARCH_LIST to select cuda archs, and this tries to:

(1) request one to explicitly pass TORCH_CUDA_ARCH_LIST as a cmake variable instead of env variable. We can make it a fatal error in the future.
(2) request one to deprecate the use of CUDA_ARCH_NAME in favor of TORCH_CUDA_ARCH_LIST.

The reason is that CUDA_ARCH_LIST is the name used in FindCUDA.cmake nowadays so it is better to consolidate.

The caffe2_* functions were originally written by NVidia circa 2014 and similar functionalities have hence been in the official cmake lists, making it better to use the official one.

Test plan: waiting for CI. Do not have cuda on my laptop.

### CODE REVIEW

File: cmake/public/cuda.cmake [hunk 1]
```diff
@@ -510,9 +323,26 @@ elseif (CUDA_VERSION VERSION_EQUAL 8.0)
 endif()
 
 # setting nvcc arch flags
-caffe2_select_nvcc_arch_flags(NVCC_FLAGS_EXTRA)
+if ((NOT EXISTS ${TORCH_CUDA_ARCH_LIST}) AND (DEFINED ENV{TORCH_CUDA_ARCH_LIST}))
+  message(WARNING
+      "In the future we will require one to explicitly pass "
```
Comment (soumith) [inline cmake/public/cuda.cmake]: from the user's commands, i.e. `python setup.py install`, they do not specify cmake variables (or dont have a straight-forward way to do so).

We can probably make setup.py to read the environment variable TORCH_CUDA_ARCH_LIST and then pass it in to the cmake manually.

Comment (Yangqing) [reply cmake/public/cuda.cmake]: Yep. setup.py shall take on the task of passing all proper env variables to cmake - and from the user front they can continue using the env behavior if needed.

**Comment by Yangqing** [comment]
Note: Windows build failure due to Eigen which is being discussed here: http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1526

**Comment by Yangqing** [comment]
@bddppq  do you mind taking a look (seems that #6746 is similar issue?)

File: cmake/public/cuda.cmake [hunk 2]
```diff
@@ -510,9 +323,26 @@ elseif (CUDA_VERSION VERSION_EQUAL 8.0)
...
+      "TORCH_CUDA_ARCH_LIST to cmake instead of implicitly setting it as an "
+      "env variable. This will become a FATAL_ERROR in future version of "
+      "pytorch.")
+  set(TORCH_CUDA_ARCH_LIST $ENV{TORCH_CUDA_ARCH_LIST})
+endif()
+if (EXISTS {CUDA_ARCH_NAME})
```
Comment (bddppq) [inline cmake/public/cuda.cmake]: missed a `$`

Comment (Yangqing) [reply cmake/public/cuda.cmake]: :( my coding skills are becoming rusted...

**Review by bddppq — APPROVED** [review]

**Comment by Yangqing** [comment]
For future reference, fix for #8269 is landed in #8271 .

**Pull request merged**